internal/cmd/commands/database/migrate.go

```diff
@@ -0,0 +1,311 @@
+	})
+
+	f.StringVar(&base.StringVar{
+		Name:   "migration-url",
+		Target: &c.flagMigrationUrl,
+		Usage:  `If set, overrides a migration URL set in config, and specifies the URL used to connect to the database for migration. This can allow different permissions for the user running initialization or migration vs. normal operation. This can refer to a file on disk (file://) from which a URL will be read; an env var (env://) from which the URL will be read; or a direct database URL.`,
+	})
+
+	return set
+}
+
+func (c *MigrateCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictNothing
+}
+
+func (c *MigrateCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *MigrateCommand) Run(args []string) (retCode int) {
+	if result := c.ParseFlagsAndConfig(args); result > 0 {
+		return result
+	}
+
+	if c.configWrapper != nil {
+		defer func() {
+			if err := c.configWrapper.Finalize(c.Context); err != nil {
+				c.UI.Warn(fmt.Errorf("Error finalizing config kms: %w", err).Error())
+			}
+		}()
+	}
+
+	dialect := "postgres"
+
+	if schema.DevMigration(dialect) != c.flagAllowDevMigrations {
+		if schema.DevMigration(dialect) {
+			c.UI.Error(base.WrapAtLength("This version of the binary has " +
+				"dev database schema updates which may not be supported in the " +
+				"next official release. To proceed anyways please use the " +
+				"'-allow-development-migrations' flag."))
+			return 2
+		} else {
+			c.UI.Error(base.WrapAtLength("The '-allow-development-migrations' " +
+				"flag was set but this binary has no dev database schema updates."))
+			return 3
+		}
+	}
+
+	c.srv = base.NewServer(&base.Command{UI: c.UI})
+
+	if err := c.srv.SetupLogging(c.flagLogLevel, c.flagLogFormat, c.Config.LogLevel, c.Config.LogFormat); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	// If mlockall(2) isn't supported, show a warning. We disable this in dev
+	// because it is quite scary to see when first using Boundary. We also disable
+	// this if the user has explicitly disabled mlock in configuration.
+	if !c.Config.DisableMlock && !mlock.Supported() {
+		c.UI.Warn(base.WrapAtLength(
+			"WARNING! mlock is not supported on this system! An mlockall(2)-like " +
+				"syscall to prevent memory from being swapped to disk is not " +
+				"supported on this system. For better security, only run Boundary on " +
+				"systems where this call is supported. If you are running Boundary" +
+				"in a Docker container, provide the IPC_LOCK cap to the container."))
+	}
+
+	if c.Config.Controller == nil {
+		c.UI.Error(`"controller" config block not found`)
+		return 1
+	}
+
+	if c.Config.Controller.Database == nil {
+		c.UI.Error(`"controller.database" config block not found`)
+		return 1
+	}
+
+	urlToParse := c.Config.Controller.Database.Url
+	if urlToParse == "" {
```

Comment:
We actually don't need this if a migration URL is set. (Technically in `init` if a migration URL is set, we don't need this either if they also skip auto creation of all initial resources.)

We could leave it in as an error, but I could imagine a use-case where someone has a config that _only_ has a migration URL so they can do upgrades but don't have access to credentials that let them modify actual data outside of the migrations. Feels nicer to me to take it out.